### ISSUE TITLE
Feature Request: apropos

### ISSUE THREAD

**Issue by dleslie** [justinethier/cyclone#96]
Unless I missed it in my cursory scan of the project, it appears that Cyclone lacks `apropos` support. (Something I wish was in RnRS). In order for [geiser](https://github.com/jaor/geiser) to support Cyclone fully it will need `apropos` support.

**Comment by justinethier** [comment]
No, there is no explicit `apropos` support. But what exactly do you need? Is geiser just going to execute functions in the REPL? It is possible for Cyclone to extract all of an environment's symbols.

**Comment by ofosos** [comment]
Hmm, the Chicken apropos facility requires among other things charset and irregex support.

@justinethier it is my understanding that we need to inspect the global symbol table to make this work. However, the symbol_table is currently not a public varialbe and the functions that are in runtime.h are not sufficient to enumerate the symbol table. What would be the best way to do this?

**Comment by justinethier** [comment]
There is a new function `env:all-variables` on master that can be used to get current variables within an environment, eg from the REPL:

    (env:all-variables *icyc-env*) 

For built-in functions, it may be possible to use static tables of symbols, which is what the chicken program seems to be doing: https://github.com/jaor/geiser/blob/master/scheme/chicken/geiser/emacs.scm#L118

**Comment by dleslie** [comment]
It needn't use irregex; the apropos utility in Chicken provides that as a user interface improvement to the API. For Geiser, I simply pass it an empty string and so extract the entire symbol table *and signatures*.

Having the symbol table is helpful for tab-completion (company-mode, auto-complete, builtin); but the signatures are necessary to support el-doc. That is, in the status-bar it will show the signature of the function at the current scope.

Off the top of my head, some other things Geiser can do if the Scheme supports it:

- Rebinding symbols within their proper module/library scope. IE, I have an R7RS library called 'foo' that exports 'bar'; and if Geiser thinks I am editing in the scope of 'foo' then it will tell the host scheme to rebind 'bar' within 'foo' and re-export it.
- Jump-to-source for the symbol-at-point or a symbol name (Chicken can't do this, but Guile can).
- Show documentation for a module, symbol, or symbol-at-point.

**Comment by ofosos** [comment]
@justinethier do you think it is possible to open up eval.sld somewhat more and expose these details? Will that work?

**Comment by justinethier** [comment]
@ofosos I think it should be possible to get symbols per my previous comment, but signatures will require more work. The compiler does not preserve any of this information, except the number of arguments.

**Comment by justinethier** [comment]
No plans to support this unless there is a spec or *something* that defines what is required here for support.
